### ISSUE TITLE
System.Threading.Tasks.Extensions with net462 pulls nearly 100 dependencies instead of 1

### ISSUE THREAD

**Issue by petedavis** [dotnet/corefx#32276]
Using Visual Studio 2017 (15.8.4)

Trying to upgrade to System.Reactive 4.1, which takes a dependency on System.Threading.Tasks.Extensions 4.5.1 on a solution using framework 4.6.2, I am getting all files from `C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Microsoft\Microsoft.NET.Build.Extensions\net461\lib` copied into my build output folder.

This is because when installed the hint path is:
`<HintPath>..\packages\System.Threading.Tasks.Extensions.4.5.1\lib\netstandard2.0\System.Threading.Tasks.Extensions.dll</HintPath>`

and not:
`<HintPath>..\packages\System.Threading.Tasks.Extensions.4.5.1\lib\portable-net45+win8+wp8+wpa81\System.Threading.Tasks.Extensions.dll</HintPath>`

Given that the package states that for net 4.5+ the only dependency should be System.Runtime.CompilerServices.Unsafe, and that package is listed as having no dependencies, this is not an expected result.

This is simple to reproduce. Create a new Net Framework project targeting framework 4.6.2. Install System.Threading.Tasks.Extensions version 4.3.0 and there will only be two dll's in the build folder.

Update to 4.4.0+ and there will be over 100 files for a package that states it has a single dependency.

The main issue now is that you can control the reference by using the mentioned workaround to point to the net45 version by manually editing the hint path. However if you migrate to package references, you can no longer control this. Which is the case for many of our project files.

**Comment by karelz** [comment]
@joperezr is it something you can help with? Is it something you want to track on GitHub, or is it unrelated to the code here and should be moved to VS Developer Feedback?

I wonder if .NET Standard 2.0 pack is dragged in somehow in the scenario.

**Comment by joperezr** [comment]
This is by design. Very similar to https://github.com/Microsoft/dotnet/blob/master/releases/net471/KnownIssues/514195-Targeting%20.NET%20Framework%204.7.1%20copies%20extra%20files%20to%20your%20bin%20directory.md except that we need more than 12 extra files since .NET 4.6.2 requires extra shims for supporting netstandard assemblies loading correctly at runtime. 

To give you a bit of context, if your project, or one of its dependencies targets netstandard, and your app is not targeting .NET Framework 4.7.2 or higher (where netstandard support comes inbox) you will need some extra assemblies in order for those netstandard assemblies to run correctly. The reason for this is that given that those assemblies are targeting a different framework than your app, the type universe doesn't fully match. What this means, is that types aren't defined on the same assembly, for example, your netstandard dependency will think that the class `String` is located at `System.Runtime.dll` while your app will think that this class is located at "mscorlib.dll" instead. Without these extra files we copy to your output directory, you will start hitting a bunch of issues of "types not found" or "methods not found". If you want to get rid of those extra files, target .NET 4.7.2 where all of them come as part of the runtime so your app won't need to carry them.

**Comment by joperezr** [comment]
I'm closing the issue since this is by-design, but feel free to reopen in case I misunderstood correctly the problem here.

**Comment by joperezr** [comment]
you have to be very careful when trying to do the "picking" logic for NuGet manually. It may be that because of the limited API that you consume from CompilerServices.Unsafe and Tasks.Extensions that your app runs fine. It is true that sometimes more than one assembly in a package could work on your app, but NuGet mappings are put in place so that the configuration that best matches your TFM is the one that is picked every time. This means that NuGet knows that the netstandard 2.0 implementation matches better what you want based on what you are targeting, so changing that logic manually is dangerous.

**Comment by i-sinister** [comment]
@joperezr, could you plz explain why adding System.Threading.Tasks.Extensions package to console project targeting 4.5/4.5.1/4.5.2 does not require "extra shims", but if projects target changed to 4.6/4.6.1/4.6.2 all of a sudden all of them are required?

Steps to reproduce (at least in VS2017)
1. Create console project "NetStandardTest"
2. Change target framework to 4.5/4.5.1/4.5.2
3. Add System.Threading.Tasks.Extensions package
4. Build the project and observe output directory:
> NetStandardTest.exe
> NetStandardTest.exe.config
> NetStandardTest.pdb
> System.Runtime.CompilerServices.Unsafe.dll
> System.Runtime.CompilerServices.Unsafe.xml
> System.Threading.Tasks.Extensions.dll
> System.Threading.Tasks.Extensions.xml

5. Create another console project that targets 4.6/4.6.1/4.6.2 or change current project target framework to 4.6/4.6.1/4.6.2. If changing framework of existing project you'll need to reinstall packages with ```Update-Package -reinstall```
6. Build project and observe output directory:
> Microsoft.Win32.Primitives.dll
> netstandard.dll
> NetStandardTest.exe
> NetStandardTest.exe.config
> NetStandardTest.pdb
> System.AppContext.dll
> System.Collections.Concurrent.dll
> System.Collections.dll
> System.Collections.NonGeneric.dll
> System.Collections.Specialized.dll
> System.ComponentModel.dll
> System.ComponentModel.EventBasedAsync.dll
> System.ComponentModel.Primitives.dll
> System.ComponentModel.TypeConverter.dll
> System.Console.dll
> System.Data.Common.dll
> System.Diagnostics.Contracts.dll
> System.Diagnostics.Debug.dll
> System.Diagnostics.FileVersionInfo.dll
> System.Diagnostics.Process.dll
> System.Diagnostics.StackTrace.dll
> System.Diagnostics.TextWriterTraceListener.dll
> System.Diagnostics.Tools.dll
> System.Diagnostics.TraceSource.dll
> System.Diagnostics.Tracing.dll
> System.Drawing.Primitives.dll
> System.Dynamic.Runtime.dll
> System.Globalization.Calendars.dll
> System.Globalization.dll
> System.Globalization.Extensions.dll
> System.IO.Compression.dll
> System.IO.Compression.ZipFile.dll
> System.IO.dll
> System.IO.FileSystem.dll
> System.IO.FileSystem.DriveInfo.dll
> System.IO.FileSystem.Primitives.dll
> System.IO.FileSystem.Watcher.dll
> System.IO.IsolatedStorage.dll
> System.IO.MemoryMappedFiles.dll
> System.IO.Pipes.dll
> System.IO.UnmanagedMemoryStream.dll
> System.Linq.dll
> System.Linq.Expressions.dll
> System.Linq.Parallel.dll
> System.Linq.Queryable.dll
> System.Net.Http.dll
> System.Net.NameResolution.dll
> System.Net.NetworkInformation.dll
> System.Net.Ping.dll
> System.Net.Primitives.dll
> System.Net.Requests.dll
> System.Net.Security.dll
> System.Net.Sockets.dll
> System.Net.WebHeaderCollection.dll
> System.Net.WebSockets.Client.dll
> System.Net.WebSockets.dll
> System.ObjectModel.dll
> System.Reflection.dll
> System.Reflection.Extensions.dll
> System.Reflection.Primitives.dll
> System.Resources.Reader.dll
> System.Resources.ResourceManager.dll
> System.Resources.Writer.dll
> System.Runtime.CompilerServices.Unsafe.dll
> System.Runtime.CompilerServices.Unsafe.xml
> System.Runtime.CompilerServices.VisualC.dll
> System.Runtime.dll
> System.Runtime.Extensions.dll
> System.Runtime.Handles.dll
> System.Runtime.InteropServices.dll
> System.Runtime.InteropServices.RuntimeInformation.dll
> System.Runtime.Numerics.dll
> System.Runtime.Serialization.Formatters.dll
> System.Runtime.Serialization.Json.dll
> System.Runtime.Serialization.Primitives.dll
> System.Runtime.Serialization.Xml.dll
> System.Security.Claims.dll
> System.Security.Cryptography.Algorithms.dll
> System.Security.Cryptography.Csp.dll
> System.Security.Cryptography.Encoding.dll
> System.Security.Cryptography.Primitives.dll
> System.Security.Cryptography.X509Certificates.dll
> System.Security.Principal.dll
> System.Security.SecureString.dll
> System.Text.Encoding.dll
> System.Text.Encoding.Extensions.dll
> System.Text.RegularExpressions.dll
> System.Threading.dll
> System.Threading.Overlapped.dll
> System.Threading.Tasks.dll
> System.Threading.Tasks.Extensions.dll
> System.Threading.Tasks.Extensions.xml
> System.Threading.Tasks.Parallel.dll
> System.Threading.Thread.dll
> System.Threading.ThreadPool.dll
> System.Threading.Timer.dll
> System.ValueTuple.dll
> System.Xml.ReaderWriter.dll
> System.Xml.XDocument.dll
> System.Xml.XmlDocument.dll
> System.Xml.XmlSerializer.dll
> System.Xml.XPath.dll
> System.Xml.XPath.XDocument.dll

Can this be caused by System.Threading.Tasks.Extensions not targeting 4.6/4.6.1/4.6.2 in the nuget package as described [here](https://weblog.west-wind.com/posts/2019/Feb/19/Using-NET-Standard-with-Full-Framework-NET)?

**Comment by joperezr** [comment]
The reason why this is happening is because you get different assets from the package when you target < 4.6 than the ones you get when you target >= 4.6. If you target < 4.6 you would get the portable asset: `lib\portable-net45+win8+wp8+wpa81\System.Threading.Tasks.Extensions.dll` which won't require all of the .net standard facades given it is itself not targeting .net standard. If you target higher than net46, then you will get the asset at `lib\netstandard2.0\System.Threading.Tasks.Extensions.dll` which depends on netstandard.dll and needs the rest of the facades to get injected in order to work properly. 

> Can this be caused by System.Threading.Tasks.Extensions not targeting 4.6/4.6.1/4.6.2 in the nuget package as described here?

Yes, if we were also providing a net461 asset in the package then you wouldn't require all of the facades to get brought in. If you think that would be valuable, I would suggest for you to file a new servicing issue on this repo to ask for System.Threading.Tasks.Extensions to add a netfx build configuration which would fix this.

**Comment by i-sinister** [comment]
@joperezr, thank you for the reply, but there'll be one more question then. As you confirmed this happens because there are no "different assets" for 4.6/4.6.1/4.6.2 and when nuget decides which "assets" to use it picks netstandard2.0 due to it is considered "newer" then 4.5, I guess. So the question is why there are no "different assets" for 4.6/4.6.1/4.6.2 just like there are for 4.5? Is there a technical limitation or a downside of adding such "assets"? Maybe it will break 'pure' netstandard2.0 libraries that use System.Threading.Tasks.Extensions?

Because I can not see how is it beneficial to pull 97 libraries with total size about 3Mb for a single library of 33Kb containing 2 interfaces, 3 enums, 6 classes and 8 structs. On the other hand I am aware of a problems caused by having netstandar2.0 being pulled into 4.6/4.6.1/4.6.2 targeting projects. Even [Microsoft itself says that](https://docs.microsoft.com/en-us/dotnet/standard/net-standard):
> While NuGet considers .NET Framework 4.6.1 as supporting .NET Standard 1.5 through 2.0, there are several issues with consuming .NET Standard libraries that were built for those versions from .NET Framework 4.6.1 projects. For .NET Framework projects that need to use such libraries, we recommend that you upgrade the project to target .NET Framework 4.7.2 or higher.

Which means "dont use netstandard2.0 in 4.6/4.6.1/4.6.2 projects". And as you may guess upgrading projects to 4.7.2 is usually/always not feasible due to deployment issues (clients just dont have 4.7 installed). So not having "different assets" for 4.6/4.6.1/4.6.2 does not sound like a good idea. Please correct me if I'm wrong.

**Comment by joperezr** [comment]
> it picks netstandard2.0 due to it is considered "newer" then 4.5, I guess.

Not really, it doesn't pick it because its newer, it picks it because netstandard2.0 is supported in 4.6.1 and it is a "closer" or "better" match than a portable assembly. If there were a pure net45 asset (instead of a portable net45 asset) then this one would have been picked as it is even a better match than netstandard2.0. 

> why there are no "different assets" for 4.6/4.6.1/4.6.2 just like there are for 4.5? Is there a technical limitation or a downside of adding such "assets"?

It wouldn't break people targeting netstandard 2.0 because we wouldn't remove the netstandard2.0 asset, we would just provide an additional asset for 461. The reason why there is no 4.6.1 asset in the package today, is because it is technically not needed since the netstandard2.0 asset is compatible with 4.6.1 as long as you carry the facades with your app. There are other internal reasons as well, since adding that asset would make our build infrastructure logic have to include this extra configuration, and it would also make the package size larger. What I mean by all this is that technically speaking, there are some small downsides on having to include a net461 asset in the package, but AFAIK there wouldn't be a breaking change in starting to include one, which is why I suggested logging a new servicing issue requesting this to be done. Once the right people triage it, they will take problems like this into consideration and then decide whether or not to service the package.

> So not having "different assets" for 4.6/4.6.1/4.6.2 does not sound like a good idea. Please correct me if I'm wrong.

This only really applies for apps that were not built with the latest tools (meaning latest versions of 2017 or 2019) since they would not include all of the required facades into the bin folder. Once you use the latest tools (which looks like in your case you are, hence all of the extra shims getting copied to your bin folder) then netstandard2.0 version of System.Threading.Tasks.Extensions.dll should run just fine on all .NET 4.6.1 machines.

**Comment by i-sinister** [comment]
> Once you use the latest tools ... then netstandard2.0 version of System.Threading.Tasks.Extensions.dll should run just fine on all .NET 4.6.1 machines.

That is not really that much simple - System.Threading.Tasks.Extensions itself is not the problem but pulling netstandard libraries which have higher versions then their counterparts in 4.6/4.6.1/4.6.2 does break the code. Most sinister one is the System.Net.Http because it is so ubiquitous and causes lots of funny "type loader", "method not found", "method does not have implementation" kind of errors if machine does not have 4.7 installed. There are workarounds for these but again - [microsoft recommends to not use .netstandard libraries in 4.6/4.6.1/4.6.2 projects](https://docs.microsoft.com/en-us/dotnet/standard/net-standard).

>  logging a new servicing issue requesting this to be done

What is 'servicing issue' and why this issue is no good? In the description it does not suggest to add target for 4.6/4.6.1/4.6.2 but it describes effects of not having one. So "fixing" this issue would be adding targets to the package (if possible).

**Comment by joperezr** [comment]
>  why this issue is no good?

That's a good point, we can use this same issue to track that work. The reason why I thought it would be better to open a new one was so that the request is more specific but I guess that is not really necessary. I'll reopen this issue and tag it for servicing (by that I mean that we have to set the milestone to 2.1.x since the System.Threading.Tasks.Extensions package is not built in corefx master any longer, so we would have to fix this via servicing in the release/2.1 branch.

**Comment by TylerBrinkley** [comment]
I was going to change [Enums.NET](https://github.com/TylerBrinkley/Enums.NET) from using a Fody implemented version of `System.Runtime.CompilerServices.Unsafe` to the nuget package but since I'm targeting .NET 4.5 it is resulting in adding a ton of .dll's to consuming code that targets the .NET Framework. Is it possible for the `System.Runtime.CompilerServices.Unsafe` package to multi-target .NET 4.5 as well?

**Comment by TylerBrinkley** [comment]
It currently targets netstandard1.0, netstandard2.0, and netcoreapp2.0, the library is implemented in IL directly so I don't think it would hold us back.

**Comment by karelz** [comment]
Even in IL you can depend on new APIs.
I would personally decide based on upvotes on such request. @ericstj any thoughts?

**Comment by TylerBrinkley** [comment]
You're right but it is already targeting .NET Standard 1.0 which is a much more limited API surface layer than .NET Framework 4.5.

**Comment by ericstj** [comment]
Easy for Unsafe, since we still build the package in master: https://github.com/dotnet/corefx/pull/42297.

For the others this is a similar change, but in servicing.

**Comment by karelz** [comment]
@ericstj are we ok with contribution for Unsafe in master?

**Comment by ericstj** [comment]
> are we ok with contribution for Unsafe in master?

Yeah, which is why I did it :).  Of course it won't ship until next year.  Is that what you are asking?  If so then that should be directed towards the folks who need this.  If they need it sooner we'd need to push it into servicing

**Comment by TylerBrinkley** [comment]
I have a workaround using Fody so I don't necessarily need this earlier and since I'm releasing a new major version of my library imminently I'd rather not wait nor pressure you to release within that time-frame. I can just switch over in my next major version after this one.

**Comment by TylerBrinkley** [comment]
@ericstj I just saw that `System.Runtime.CompilerServices.Unsafe` v4.7.0 was released but it still doesn't target .NET Framework 4.5. Was that intentional? Will your change be included in the next version?

**Comment by ericstj** [comment]
Right, 4.7.0 was for the 3.1 release, but this change was made to master which is 5.0.

**Comment by TylerBrinkley** [comment]
Perfect, thank you.
